saleor/graphql/product/dataloaders/products.py

```diff
@@ -209,47 +211,59 @@ class ProductVariantsByProductIdAndChannel(
     context_key = "productvariant_by_product_and_channel"
 
     def batch_load(self, keys: Iterable[tuple[int, str]]):
-        product_ids, channel_slugs = zip(*keys)
-        variants_filter = self.get_variants_filter(product_ids, channel_slugs)
-
-        variants = (
-            ProductVariant.objects.using(self.database_connection_name)
-            .filter(**variants_filter)
-            .annotate(channel_slug=F("channel_listings__channel__slug"))
-            .order_by("sort_order", "sku")
+        _, channel_slugs = zip(*keys)
+
+        product_ids_by_channel = defaultdict(list)
+        for product_id, channel_slug in keys:
+            product_ids_by_channel[channel_slug].append(product_id)
+
+        def with_channels(channels):
+            channel_map = {c.slug: c for c in channels}
+            variant_map: defaultdict[
+                tuple[int, str], list[ProductVariant]
+            ] = defaultdict(list)
+
+            for channel_slug in product_ids_by_channel.keys():
+                channel = channel_map[channel_slug]
+                product_ids = product_ids_by_channel[channel_slug]
+                variants_filter = self.get_variants_filter(channel.id)
+
+                variants = (
+                    ProductVariant.objects.using(self.database_connection_name)
+                    .filter(variants_filter)
+                    .filter(product_id__in=product_ids)
+                    .order_by("sort_order", "sku")
+                )
+
+                for variant in variants.iterator():
+                    key = (variant.product_id, channel.slug)
+                    variant_map[key].append(variant)
+
+            return [variant_map.get(key, []) for key in keys]
+
+        return (
+            ChannelBySlugLoader(self.context)
+            .load_many(channel_slugs)
```

Comment:
How about remove line 214('_, channel_slugs = zip(*keys)')?
```suggestion
            .load_many(product_ids_by_channel.keys())
```